backend/src/main/java/com/bakdata/conquery/sql/conversion/model/select/SumDistinctSqlAggregator.java

```diff
@@ -0,0 +1,199 @@
+
+		private final String suffix;
+		private final SumDistinctCteStep predecessor;
+
+		SumDistinctCteStep(String suffix, SumDistinctCteStep predecessor) {
+			this.suffix = suffix;
+			this.predecessor = predecessor;
+		}
+
+		@Override
+		public String cteName(String nodeLabel) {
+			return "%s-%s".formatted(nodeLabel, this.suffix);
+		}
+
+		@Override
+		public CteStep predecessor() {
+			return this.predecessor;
+		}
+	}
+
+	private static final String ROW_NUMBER_ALIAS = "row_number";
+	private static final String SUM_DISTINCT_SUFFIX = "sum_distinct";
+
+	SqlSelects sqlSelects;
+	Filters filters;
+
+	public SumDistinctSqlAggregator(
+			Column sumColumn,
+			List<Column> distinctByColumns,
+			String alias,
+			IRange<? extends Number, ?> filterValue,
+			Field<Object> primaryColumn,
+			SqlTables<ConceptCteStep> conceptTables,
+			NameGenerator nameGenerator
+	) {
+		String rootTable = conceptTables.getRootTable();
+		Class<? extends Number> numberClass1 = NumberMapUtil.NUMBER_MAP.get(sumColumn.getType());
+		ExtractingSqlSelect<? extends Number> sumColumnRootSelect = new ExtractingSqlSelect<>(rootTable, sumColumn.getName(), numberClass1);
+		List<ExtractingSqlSelect<Object>> distinctByRootSelects = distinctByColumns.stream()
+																				   .map(column -> new ExtractingSqlSelect<>(
+																						   rootTable, column.getName(), Object.class)
+																				   )
+																				   .toList();
+
+		QueryStep rowNumberCte = createRowNumberCte(primaryColumn, sumColumnRootSelect, distinctByRootSelects, alias, conceptTables, nameGenerator);
+		FieldWrapper<BigDecimal> sumSelect = createSumDistinctSelect(sumColumnRootSelect, alias, rowNumberCte);
+		QueryStep rowNumberFilteredCte = createRowNumberFilteredCte(rowNumberCte, primaryColumn, sumSelect, alias, nameGenerator);
+
+		SqlSelects.SqlSelectsBuilder builder = SqlSelects.builder()
+														 .preprocessingSelect(sumColumnRootSelect)
+														 .preprocessingSelects(distinctByRootSelects)
+														 .additionalPredecessor(rowNumberFilteredCte);
+
+		if (filterValue != null) {
+			this.sqlSelects = builder.build();
+			Field<BigDecimal> qualifiedSumSelect = sumSelect.createAliasedReference(conceptTables.getPredecessor(ConceptCteStep.AGGREGATION_FILTER)).select();
+			SumCondition sumCondition = new SumCondition(qualifiedSumSelect, filterValue);
+			this.filters = Filters.builder()
+								  .group(List.of(sumCondition))
+								  .build();
+		}
+		else {
+			ExtractingSqlSelect<BigDecimal> finalSelect = sumSelect.createAliasedReference(conceptTables.getPredecessor(ConceptCteStep.FINAL));
+			this.sqlSelects = builder.finalSelect(finalSelect).build();
+			this.filters = Filters.builder().build();
+		}
+	}
+
+	/**
+	 * Assigns row numbers for each partition over the pid and the distinct by columns. If the values per pid in the distinct by columns are duplicated,
+	 * the row number will be incremented for each duplicated entry.
+	 */
```

Comment:
wie im Chat erwähnt erscheint mir der Ansatz zu kompliziert

Reply:
Ich hab mir den Vorschlag angeschaut und vermute mal, dass du dir das so vorgestellt hast:
```sql
with "grouped_by_distinct_columns" as (select "pid", (array_agg("value"))[1] as "value"
                                       from "distinctOnTable"
                                       group by "pid", "k1", "k2")
select "pid", sum("value")
from "grouped_by_distinct_columns"
group by "pid";
```

Das habe ich jetzt so umgeändert, allerdings bleibt es beim grundlegenden Aufbau mit den zusätzlichen CTEs. Ob ich eine dann eine zusätzliche CTE nutze oder eine Subquery läuft ja auf dasselbe Ergebnis hinaus. Mir (uns) erschien es deutlich besser lesbar und verständlicher, die Subquery dann nicht inlined sondern als extra CTE zu haben.

```sql
[...]
"sum_distinct_select-1-grouped_by_distinct_columns" as (
        select
            "concept_concept-1-preprocessing"."pid",
            (array_agg("concept_concept-1-preprocessing"."value"))[1] as "sum_distinct_select-1"
        from "concept_concept-1-preprocessing"
        group by "concept_concept-1-preprocessing"."pid", "concept_concept-1-preprocessing"."k1", "concept_concept-1-preprocessing"."k2"
    ),
    "sum_distinct_select-1-sum_distinct" as (
        select
            "sum_distinct_select-1-grouped_by_distinct_columns"."pid",
            sum("sum_distinct_select-1-grouped_by_distinct_columns"."sum_distinct_select-1") as "sum_distinct_select-1"
        from "sum_distinct_select-1-grouped_by_distinct_columns"
        group by "sum_distinct_select-1-grouped_by_distinct_columns"."pid"
    ),
[...]
```

Hier mal ein Vergleich wie es inlined aussieht: 
```sql
[...]
"sum_distinct_select-1-sum_distinct" as (select "distinct_select"."pid",
                                                     sum("distinct_select"."sum_distinct_select-1") as "sum_distinct_select-1"
                                              from (select "concept_concept-1-preprocessing"."pid",
                                                           (array_agg("concept_concept-1-preprocessing"."value"))[1] as "sum_distinct_select-1"
                                                    from "concept_concept-1-preprocessing"
                                                    group by "concept_concept-1-preprocessing"."pid",
                                                             "concept_concept-1-preprocessing"."k1",
                                                             "concept_concept-1-preprocessing"."k2") "distinct_select"
                                              group by "distinct_select"."pid"),
[...]
```

Reply:
achso ja das war definitv mit CTE gemeint. Das Tool was ich zum testen benutzt habe konnte keine CTEs.

Dieses pattern mit `array_agg(...)[0]` erscheint mir ziemlich ineffizient. Ist FIRST/LAST nicht standardisiert in SQL?

Reply:
@awildturtok Hab nochmal nachgeschaut, first/last ist bei Postgres nur eine Window-Function, für Aggregationen gibt es nichts Vergleichbares. Der Weg mit `array_agg` wird an vielen Stellen als mögliche Lösung vorgeschlagen, sicherlich gibt es auch andere Wege.